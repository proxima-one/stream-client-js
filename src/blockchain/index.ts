export * from "./model";
export * from "./blockEventStreamer";
