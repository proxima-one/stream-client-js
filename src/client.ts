import { Stream } from "./model/stream";
import { StreamDBConsumerClient } from "./stream-db/streamDbDatabase";
import {
    StreamInfo,
    Offset, StateTransition, PausableStream, SimplePauseController
  } from "@proxima-one/proxima-streams";
import axios from "axios";


export class ProximaStreamsClient {
    public endpoint;
    constructor(endpoint = "https://stream-api.cluster.amur-dev.proxima.one") {
        this.endpoint = endpoint
    }

    async getStreams(): Promise<Stream[]> {
        try {
            const streams = await axios.get(this.endpoint + "/streams")
            return streams.data
        } catch(e) {
            console.log(e)
            return []
        }
    }

    async getStream(name: string, retries = 10): Promise<Stream | undefined> {
        try {
            const resp = await axios.get(this.endpoint + "/stream:" + name)
            if (resp.data) {
                const stream = resp.data
                return stream
            }  else if (retries > 0) {
                return this.getStream(name, retries - 1)
            } else {
                throw new Error("Cannot get stream")
            }
        } catch(e) {
            console.log(e)
            return undefined
        }
    }

    async getStreamConsumer(name: string, retries = 10): Promise<StreamDBConsumerClient> {
        try {
            const stream = await this.getStream(name)
            if (stream && stream.endpoints && stream.endpoints.length > 0) {
                return new StreamDBConsumerClient(stream.endpoints[0])
            } else if (retries > 0) {
                return this.getStreamConsumer(name, retries - 1)
            } else {
                throw new Error("Cannot get stream client")
            }
        } catch(e) {
            console.log(e)
            throw new Error("Cannot get stream client" + e)
        }
    }

    async findOffset(stream: string, height: bigint) {
        try {
            const consumer = await this.getStreamConsumer(stream)
            if (consumer) {
                return consumer.findOffset(stream, height)
            }
            return undefined
        } catch(e) {
            console.log(e)
            return undefined
        }
    }

    async getStateTransitions(stream: string, offset: Offset, count: number, direction: "next" | "last") {
        try {
            const consumer = await this.getStreamConsumer(stream)
            if (consumer) {
                return await consumer.getStateTransitions(stream, offset, count, direction)
            } 
            return undefined
        } catch(e) {
            console.log(e)
            return undefined
        }
    }

    async streamStateTransitions(stream: string, offset: Offset): Promise<PausableStream<StateTransition>> {
        try {
            const consumer = await this.getStreamConsumer(stream)
            if (consumer) {
                return await consumer.streamStateTransitions(stream, offset)
            } else {
                throw new Error("Could not get pausable stream")
            }
        } catch(e) {
            console.log(e)
            throw new Error("Could not get pausable stream" + e)
        }
    }
}