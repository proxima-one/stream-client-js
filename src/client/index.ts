export * from "./proximaStreamClient";
export * from "./streamRegistry";
export * from "./singleStreamDbRegistry";
export * from "./streamRegistryClient";
