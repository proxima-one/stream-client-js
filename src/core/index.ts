export * from "./model";
export * from "./queries";
export * from "./serialization";
export * from "./common";
export * from "./ethDataSource";
//export * from "./blockIndexer";
