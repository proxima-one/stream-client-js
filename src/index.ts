export * from "./client";
//import * as ProximaServiceTypes from "./gen/proto/messages/v1alpha1/messages_pb";
