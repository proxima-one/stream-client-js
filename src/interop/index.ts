export * from "./endpoints";
export * from "./queries";
export * from "./ethInterop";
export * from "./ethInteropConfig";
export * from "./queryStats";
export * from "./endpoints/web3EndpointFactory";
export * from "./timeProvider";
