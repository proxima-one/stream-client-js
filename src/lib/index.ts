export { PausableStream } from "./pausableStream";
export * from "./reader";
