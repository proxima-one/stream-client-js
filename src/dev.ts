import { Offset } from "./model";
import { ProximaStreamClient, SingleStreamDbRegistry, StreamRegistryClient } from "./client";
import { BufferedStreamReader } from "./lib/reader";
import { sleep } from "./utils";

async function main() {
  // connect to local StreamDB
  // const client = new ProximaStreamClient({
  //   registry: new SingleStreamDbRegistry("localhost:50051"),
  // });

  // by default connect through Stream Registry
  const client = new ProximaStreamClient();

  const streamRegistry = new StreamRegistryClient();
  const allStreams = await streamRegistry.getStreams();
  console.log(allStreams.map(x => x.name));

  let currentOffset = Offset.zero;
  for (let i = 0; i < 10; i++) {
    const events = await client.fetchEvents("proxima.eth-main.blocks.1_0", currentOffset, 100, "next");
    currentOffset = events[events.length - 1].offset;

    console.log(`fetched till ${currentOffset.toString()}`);
  }

  console.log("starting the stream");

  const stream = await client.streamEvents("proxima.eth-main.blocks.1_0", Offset.zero);
  const reader = BufferedStreamReader.fromStream(stream, 10000);

  while (true) {
    const batch = await reader.read(100);
    if (!batch)
      break;

    console.log(`doing some stuff with a batch of ${batch.length}. ${batch[batch.length-1].offset.toString()}`);
    // simulate processing
    await sleep(100);
  }
}

main();
