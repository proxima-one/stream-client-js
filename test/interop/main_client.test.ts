import { ProximaStreamsClient} from "../../src";
import { firstValueFrom, map, take, toArray } from "rxjs";

const testEndpoint = "https://stream-api.cluster.amur-dev.proxima.one";

describe("MainStreamClient", () => {
  let client: ProximaStreamsClient;
  beforeEach(() => {
    client = new ProximaStreamsClient(testEndpoint);
  });

  it("should get all streams", async () => {
    const streams =  await client.getStreams()
    expect(
      streams.map((x) =>
        x.meta
      )
    ).toMatchSnapshot();
  });

  it("should get a single stream information", async () => {
    const streams = await client.getStreams()
    const stream_name = ""
    const stream = await client.getStream(stream_name)
    //expect(stream).toBeDefined()
    //expect(JSON.parse(stream.meta.toString("utf8"))).toMatchSnapshot();
  });

  // it("should get a single stream consumer for streamdbv2", async () => {
  //   const stream_name = "polygon-mumbai-blockheader.streams.proxima.one"
  //   const streamConsumer = await client.getStreamConsumer(stream_name)
  // });

  // it("should get a single streams", async () => {
  //   const stream_name = "polygon-mumbai-blockheader.streams.proxima.one"
  //   const streamConsumer = await client.getStreamConsumer(stream_name)
  // });
});
