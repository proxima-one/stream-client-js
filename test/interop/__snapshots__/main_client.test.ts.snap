// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MainStreamClient should get all streams 1`] = `
Array [
  Object {
    "description": "",
    "labels": Object {
      "encoding": "json",
    },
  },
  Object {
    "description": "",
    "labels": Object {
      "encoding": "json",
    },
  },
  Object {
    "description": "",
    "labels": Object {
      "encoding": "json",
    },
  },
  Object {
    "description": "",
    "labels": Object {
      "encoding": "json",
    },
  },
  Object {
    "description": "",
    "labels": Object {
      "encoding": "json",
    },
  },
  Object {
    "description": "",
    "labels": Object {
      "encoding": "json",
    },
  },
  Object {
    "description": "",
    "labels": Object {
      "encoding": "json",
    },
  },
  Object {
    "description": "",
    "labels": Object {
      "encoding": "json",
    },
  },
  Object {
    "description": "",
    "labels": Object {
      "encoding": "json",
    },
  },
  Object {
    "description": "",
    "labels": Object {
      "encoding": "json",
    },
  },
  Object {
    "description": "",
    "labels": Object {
      "encoding": "json",
    },
  },
  Object {
    "labels": Object {
      "encoding": "json",
    },
  },
  Object {
    "description": "",
    "labels": Object {
      "encoding": "json",
    },
  },
  Object {
    "description": "",
    "labels": Object {
      "encoding": "json",
    },
  },
  Object {
    "description": "",
    "labels": Object {
      "encoding": "json",
    },
  },
]
`;
